File: Loenn/entities/Cornerspike.lua
local vivUtil = require('mods').requireFromPlugin('libraries.vivUtil')

local edges = {"UpLeft","UpRight","DownRight","DownLeft"}

local renderSupport = {
    UpLeft = {1,1},
    UpRight = {0,1},
    DownLeft = {1,0},
    DownRight = {0,0},
}

local cs = {
    name = "VivHelper/CornerSpike",
    fieldOrder = {"type", "Color",
        "EdgeDirection","KillFormat", "DoNotAttach"
    },
    fieldInformation = {
        EdgeDirection = {fieldType = "string", options = {
        {"Top Left", "UpLeft"},
        {"Top Right", "UpRight"},
        {"Bottom Left", "DownLeft"},
        {"Bottom Right", "DownRight"},
        {"Inner Bottom Right", "InnerUpLeft"},
        {"Inner Bottom Left", "InnerUpRight"},
        {"Inner Top Right", "InnerDownLeft"},
        {"Inner Top Left", "InnerDownRight"}
        }},
        Color = {fieldType = "VivHelper.color", allowXNAColors = true }
    },
    depth = -2,
    texture = function(room,entity)
        local t = entity.type
        t = ((t == "default" or t == "outline") and "danger/spikes/corners/" .. t or t)
        return t .. "_" .. entity.EdgeDirection .. "00"
    end,
    color = function(room,entity) 
        vivUtil.getColorTable(entity.Color, true, {1,1,1,1})
    end,
    justification = function(room,entity)
        local sub = entity.EdgeDirection:sub(entity.EdgeDirection:sub(1, 5) == "Inner" and 6 or 0)
        return renderSupport[sub]
    end

}
local function starts(String,Start)
    return string.sub(String,1,string.len(Start))==Start
 end

cs.rotate = function(room,entity,direction)
    local a = starts(entity.EdgeDirection,"Inner") and 5 or 0
    local pre = string.sub(entity.EdgeDirection, 1, a)
    local post = string.sub(entity.EdgeDirection, a+1)
    local startIndex = vivUtil.indexof(edges, post)
    if not startIndex then return false end
    local realIndex = vivUtil.mod(startIndex - 1 + direction, 4) + 1
    entity.EdgeDirection = pre .. edges[realIndex]

    return true
end
cs.flip = function(room,entity,horizontal,vertical)
    if vertical then
        entity.EdgeDirection = entity.EdgeDirection:gsub("Up","XXX"):gsub("Down","Up"):gsub("XXX","Down")
    end
    if horizontal then
        entity.EdgeDirection = entity.EdgeDirection:gsub("Left","XXX"):gsub("Right","Left"):gsub("XXX","Right")
    end
end
cs.placements = {}
for _,edge in ipairs(edges) do
    table.insert(cs.placements,{
    name = edge, data = {
        type = "default",
        Color = "ffffff",
        EdgeDirection = edge,
        DoNotAttach = true, KillFormat = false
    }})
    
end
return cs

File: Ahorn/entities/Cornerspike.jl
module VivHelperCornerspike
using ..Ahorn, Maple
using Ahorn.VivHelper

@mapdef Entity "VivHelper/CornerSpike" CornerSpike(x::Integer, y::Integer, type::String="default", Color::String="White", DoNotAttach::Bool=false)

const placements = Ahorn.PlacementDict(
    "Cornerspike (Top Left Outer) (Viv's Helper)" => Ahorn.EntityPlacement(
        CornerSpike,
        "rectangle",
        Dict{String, Any}(
            "EdgeDirection" => "UpLeft",
            "KillFormat" => false
        )
    ), 
    "Cornerspike (Top Right Outer) (Viv's Helper)" => Ahorn.EntityPlacement(
        CornerSpike,
        "rectangle",
        Dict{String, Any}(
            "EdgeDirection" => "UpRight",
            "KillFormat" => false
        )
    ),
    "Cornerspike (Bottom Left Outer) (Viv's Helper)" => Ahorn.EntityPlacement(
        CornerSpike,
        "rectangle",
        Dict{String, Any}(
            "EdgeDirection" => "DownLeft",
            "KillFormat" => false
        )
    ),
    "Cornerspike (Bottom Right Outer) (Viv's Helper)" => Ahorn.EntityPlacement(
        CornerSpike,
        "rectangle",
        Dict{String, Any}(
            "EdgeDirection" => "DownRight",
            "KillFormat" => false
        )
    ),
    "Cornerspike (Bottom Right Inner) (Viv's Helper)" => Ahorn.EntityPlacement(
        CornerSpike,
        "rectangle",
        Dict{String, Any}(
            "EdgeDirection" => "InnerUpLeft",
            "KillFormat" => false
        )
    ), 
    "Cornerspike (Bottom Left Inner) (Viv's Helper)" => Ahorn.EntityPlacement(
        CornerSpike,
        "rectangle",
        Dict{String, Any}(
            "EdgeDirection" => "InnerUpRight",
            "KillFormat" => false
        )
    ),
    "Cornerspike (Top Right Inner) (Viv's Helper)" => Ahorn.EntityPlacement(
        CornerSpike,
        "rectangle",
        Dict{String, Any}(
            "EdgeDirection" => "InnerDownLeft",
            "KillFormat" => false
        )
    ),
    "Cornerspike (Top Left Inner) (Viv's Helper)" => Ahorn.EntityPlacement(
        CornerSpike,
        "rectangle",
        Dict{String, Any}(
            "EdgeDirection" => "InnerDownRight",
            "KillFormat" => false
        )
    )
)

Ahorn.editingOptions(entity::CornerSpike) = Dict{String, Any}(
    "EdgeDirection" => Dict{String, String}(
        "Top Left" => "UpLeft",
        "Top Right" => "UpRight",
        "Bottom Left" => "DownLeft",
        "Bottom Right" => "DownRight",
        "Inner Bottom Right" => "InnerUpLeft",
        "Inner Bottom Left" => "InnerUpRight",
        "Inner Top Right" => "InnerDownLeft",
        "Inner Top Left" => "InnerDownRight"
    ),
    "Color" => VivHelper.XNAColors,
    "type" => String["default", "outline"]
)

const DirOffsets = Dict{String, Tuple{Integer, Integer}}(
    "UpLeft" => (-8, -8),
    "UpRight" => (0, -8),
    "DownLeft" => (-8, 0),
    "DownRight" => (0, 0),
    "InnerUpLeft" => (-8, -8),
    "InnerUpRight" => (0, -8),
    "InnerDownLeft" => (-8, 0),
    "InnerDownRight" => (0, 0)
)

function DirectionOffsets(entity::CornerSpike)
    a =get(entity.data, "EdgeDirection", "")
    return haskey(DirOffsets, a) ? DirOffsets[a] : (0, 0)
end

function Ahorn.selection(entity::CornerSpike)
    x, y = Ahorn.position(entity)
    a, b = DirectionOffsets(entity)
    Ahorn.Rectangle(x+a,y+b,8,8)

end

function Ahorn.renderAbs(ctx::Ahorn.Cairo.CairoContext, entity::CornerSpike)
    a = get(entity.data, "EdgeDirection", "")
    b0 = get(entity.data, "type", "default")
    x, y = Ahorn.position(entity)
    if a == "" || b0 == ""
        Ahorn.drawSprite(ctx, "", x, y, jx = 0.5, jy = 0.5)
    else
        b = string(b0 == "default" ? "danger/spikes/corners/default" : b0 == "outline" ? "danger/spikes/corners/outline" : b0, "_")
        c, d = DirectionOffsets(entity)
        oneColor = get(entity.data, "Color", "")
        oneColor = VivHelper.ColorFix((oneColor == "" ? "White" : oneColor), 1.0)
        Ahorn.drawSprite(ctx, string(b,a,"00"), x+(c * 9 / 8), y+(d * 9 / 8), jx=0, jy=0, tint=oneColor)
    end
end

end